client/offchain/src/api/timer.rs

```diff
@@ -0,0 +1,180 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This module is composed of two structs: [`TimerApi`] and [`TimerWorker`]. Calling the [`timer`]
+//! function returns a pair of [`TimerApi`] and [`TimerWorker`] that share some state.
+//!
+//! For more technical details and reasoning behind this design, see the analogous [`http`] module.
+
+use sp_core::offchain;
+use sp_core::offchain::Timestamp;
+use sp_core::offchain::PollableId;
+use sp_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+
+use core::cmp::{Ordering, Reverse};
+use core::future::Future;
+use core::pin::Pin;
+use core::task::{self, Poll};
+use core::time;
+use std::collections::BinaryHeap;
+
+use futures::Stream;
+use futures_timer::Delay;
+
+use super::timestamp;
+
+pub use sp_core::offchain::TimerId;
+
+pub fn timer(sink: TracingUnboundedSender<PollableId>) -> (TimerApi, TimerWorker) {
+	let (to_worker, from_api) = tracing_unbounded("mpsc_ocw_timer_from");
+
+	let worker = TimerWorker {
+		ready_ids: sink,
+		from_api,
+		delay: None,
+		ids: Default::default(),
+	};
+
+	let api = TimerApi {
+		to_worker,
+		next_id: TimerId(0),
+	};
+
+	(api, worker)
+}
+
+pub struct TimerApi {
+	/// Used to enqueue new timer in the `TimerWorker`.
+	to_worker: TracingUnboundedSender<(TimerId, Timestamp)>,
+	/// Counter used to generate new timer IDs.
+	next_id: TimerId,
+}
+
+impl TimerApi {
+	/// Starts a new timer that resolves a `duration` from the current epoch.
+	pub fn start_timer(&mut self, duration: offchain::Duration) -> Result<TimerId, ()> {
+		let id = self.next_id;
+		self.next_id = TimerId(self.next_id.0 + 1);
+
+		let timestamp = timestamp::now().add(duration);
+
+		self.to_worker.unbounded_send((id, timestamp))
+			.map(|_| id)
+			.map_err(drop)
+	}
+}
+
+/// A `TimerId` wrapper that implements `Ord` and `Eq` using an additional
+/// `Timestamp` value.
+struct TimerIdWithTimestamp {
+	key: Timestamp,
+	id: TimerId,
+}
+
+impl PartialEq for TimerIdWithTimestamp {
+	fn eq(&self, other: &Self) -> bool {
+		PartialEq::eq(&self.key, &other.key)
+	}
+}
+
+impl Eq for TimerIdWithTimestamp {}
+
+impl PartialOrd for TimerIdWithTimestamp {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		PartialOrd::partial_cmp(&self.key, &other.key)
+	}
+}
+
+impl Ord for TimerIdWithTimestamp {
+	fn cmp(&self, other: &Self) -> Ordering {
+		Ord::cmp(&self.key, &other.key)
+	}
+}
+
+pub struct TimerWorker {
```

Comment:
Can't we use existing `future` timers library for that? Why implement that from scratch? It's host code anyway, no?